### PR TITLE
[Coverage] Support jump expressions

### PR DESCRIPTION
**Pull request by shadrina**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt [hunk 1]
```diff
@@ -71,32 +71,57 @@ private class IrFunctionRegionsCollector(
...
 
     override fun visitConstructor(declaration: IrConstructor) {
         val statements = declaration.body?.statements ?: return
-        statements.forEach {
-            if (it is IrDelegatingConstructorCall && !declaration.isPrimary
-                    || it !is IrDelegatingConstructorCall && it !is IrReturn) {
-                recordRegion(it)
-                it.acceptVoid(this)
+        visitInStatementsContext(statements) {
+            if (this is IrDelegatingConstructorCall && !declaration.isPrimary
```
Comment (sbogolepov) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt]: Nesting of several `this` is not a good solution from reader's point of view and should be avoided whenever possible. The reason is that it requires reader to remember that this `this` is not pointing to `IrFunctionRegionsCollector` instance but rather to implicit `IrStatement` that came from `visitInStatementsContext`. More obvious code is better :)

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt [hunk 2]
```diff
@@ -117,48 +142,70 @@ private class IrFunctionRegionsCollector(
...
+        val region = Region.fromIr(irElement, currentFile) ?: return
+        regions[irElement] = region
+        regionStack.push(region)
+        visit()
+        regionStack.pop()
+    }
+
+    private fun visitInStatementsContext(
+            statements: List<IrStatement>,
+            visit: IrStatement.() -> Unit = { acceptVoid(this@IrFunctionRegionsCollector) }
```
Comment (sbogolepov) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt]: `visit: (IrStatement) -> Unit` is better as explained above.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt [hunk 3]
```diff
@@ -71,32 +71,57 @@ private class IrFunctionRegionsCollector(
         val irFile: IrFile
 ) : IrElementVisitorVoid {
 
+    val regions = mutableMapOf<IrElement, Region>()
+
     private val irFileStack = mutableListOf(irFile)
 
+    private val regionStack = mutableListOf<Region>()
+
+    private val irStatementsStack = mutableListOf<Pair<IrStatement, IrStatement>>()
```
Comment (sbogolepov) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt]: Let's replace `Pair<IrStatement, IrStatement>` with `data class` that has meaningful names for its elements.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt [hunk 4]
```diff
@@ -117,48 +144,69 @@ private class IrFunctionRegionsCollector(
...
+        visit()
+        irFileStack.pop()
     }
 
-    private fun recordRegion(irElement: IrElement, startOffset: Int, endOffset: Int, kind: RegionKind = RegionKind.Code) {
-        if (startOffset == UNDEFINED_OFFSET || endOffset == UNDEFINED_OFFSET) {
-            return
+    private fun visitInRegionContext(irElement: IrElement, visit: () -> Unit) {
+        val region = Region.fromIr(irElement, currentFile) ?: return
+        regions[irElement] = region
```
[the remark this answers is not in the document]
Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt]: BTW, It looks like that `visitInRegionContext` and `visitInFileContext` perform two separate actions:
1. Recording region for the first parameter
2. Performs `visit`
It is also inconsistent with `visitInStatementContext`.
So moving `recordRegion` out of these functions will do the work.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt [hunk 5]
```diff
@@ -117,48 +144,69 @@ private class IrFunctionRegionsCollector(
...
-                    expression.acceptChildrenVoid(this)
-                    irFileStack.pop()
-                }
-            }
-            else -> expression.acceptChildrenVoid(this)
-        }
+    override fun visitReturn(expression: IrReturn) {
+        val (_, next) = irStatementsStack.lastOrNull() ?: return
+        val nextRegion = Region.fromOffset(expression.endOffset, currentRegion.endOffset, currentFile) ?: return
+        regions[next] = nextRegion
```
[the remark this answers is not in the document]
Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt]:
```
fun main() {
    return
}
```
It was fully covered before, but now the last line is not covered.

[the remark this answers is not in the document]
Comment (shadrina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt]: It turns out, there is an additional `IrReturn`-statement at the end of `IrFunction` with an incorrect empty region. So I forced to add offsets validity check before visiting the statement inside `visitInStatementContext`: [click](https://github.com/JetBrains/kotlin-native/pull/3630/files#diff-afed526d846ac36c39d0e76b9a44deb6R199). 

Anyway, added the sample to the test, it works now as expected.

**Pull request merged**